src/binary_descriptors/mod.rs

```diff
@@ -0,0 +1,170 @@
+//! Functions for generating and comparing compact binary patch descriptors.
+
+use rand::Rng;
+use std::collections::HashMap;
+
+use crate::point::Point;
+
+pub mod brief;
+mod constants;
+
+/// A feature descriptor whose value is given by a string of bits.
+pub trait BinaryDescriptor {
+    /// Returns the length of the descriptor in bits. Typical values are 128,
+    /// 256, and 512. Will always be a multiple of 128.
+    fn get_size(&self) -> u32;
+    /// Returns the number of bits that are different between the two descriptors.
+    ///
+    /// Panics if the two descriptors have unequal lengths. The descriptors
+    /// should have been computed using the same set of test pairs, otherwise
+    /// comparing them has no meaning.
+    fn compute_hamming_distance(&self, other: &Self) -> u32;
+    /// Given a set of bit indices, returns those bits from the descriptor as a
+    /// single concatenated value.
+    ///
+    /// Panics if `bits.len() > 128`.
+    fn get_bit_subset(&self, bits: &[u32]) -> u128;
+    /// Returns the pixel location of this binary descriptor in its associated
+    /// image.
+    fn get_position(&self) -> Point<u32>;
+}
+
+// impl Display for BriefDescriptor {
+//     fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+//         for chunk in self.bits.iter() {
+//             write!(f, "{:#32x}", chunk)?;
+//         }
+//         Ok(())
+//     }
+// }
+
+/// For each descriptor in `d1`, find the descriptor in `d2` with the minimum
+/// Hamming distance below `threshold`. If no such descriptor exists in `d2`,
+/// the descriptor in `d1` is left unmatched.
+///
+/// Descriptors in `d2` may be matched with more than one descriptor in `d1`.
+///
+/// Uses [locality-sensitive hashing][lsh] (LSH) for efficient matching. The
+/// number of tables is fixed at three, but the hash length is proportional to
+/// the log of the size of the largest input array.
+///
+/// Returns a vector of references describing the matched pairs. The first
+/// reference is to a descriptor in `d1`, and the second reference is to an
+/// index into `d2`.
+///
+/// [lsh]:
+///     https://en.wikipedia.org/wiki/Locality_sensitive_hashing#Bit_sampling_for_Hamming_distance
+pub fn match_binary_descriptors<'a, T: BinaryDescriptor>(
+    d1: &'a [T],
+    d2: &'a [T],
+    threshold: u32,
+) -> Vec<(&'a T, &'a T)> {
+    // early return if either input is empty
+    if d1.is_empty() || d2.is_empty() {
+        return Vec::new();
+    }
+
+    let mut rng = rand::thread_rng();
```

Comment:
It would be nice to allow deterministic reproduction for e.g. regression testing. This could take the rng as in input or take an optional seed? Or non-deterministic is fine.